syntax = "proto2";

package util;

option java_package = "bftsmart.proto";

option java_multiple_files = true;

option cc_generic_services = true;

message RPCRequest {
  required int32 localId = 1;
  optional int32 sequence = 2;
  required bytes payload = 4;
};

message DeliverRequest {
  required int32 localId = 1;
  optional int32 sequence = 2;
  optional int32 leaderId = 3;
  repeated bytes contents = 4;
  repeated bytes signatures = 5;
};

message RPCResponse {
  required bool success = 1;
  optional bytes payload = 2;
};

service RPCService {
  // call by the leader, RPCResponse return the payload
  rpc requestProposal(RPCRequest) returns (RPCResponse);
  rpc signProposal(RPCRequest) returns (RPCResponse);
  rpc verifyProposal(RPCRequest) returns (RPCResponse);
  rpc deliver(DeliverRequest) returns (RPCResponse);
  rpc leaderStart(RPCRequest) returns (RPCResponse);
  // request of leaderStop contains the new leader
  rpc leaderStop(RPCRequest) returns (RPCResponse);
};
